avatars/animationHelpers.js

```diff
@@ -412,1069 +423,789 @@ export const loadPromise = (async () => {
+    //
+    avatar.landNodeTwoPtr = avatar.createNode(AnimationNodeType.TWO, 'landNodeTwoPtr');
+    physx.physxWorker.addChild(avatar.landNodeTwoPtr, avatar.fallLoopNodeTwoPtr);
+    physx.physxWorker.addChild(avatar.landNodeTwoPtr, avatar.landsNodeSolitaryPtr);
+
+    avatar.activatesNodeSolitaryPtr = avatar.createNode(AnimationNodeType.SOLITARY, 'activatesNodeSolitaryPtr');
+    for (const k in avatar.activateMotionPtro) {
+      const motion = avatar.activateMotionPtro[k];
+      physx.physxWorker.addChild(avatar.activatesNodeSolitaryPtr, motion);
+    }
+    avatar.activateNodeTwoPtr = avatar.createNode(AnimationNodeType.TWO, 'activateNodeTwoPtr');
+    physx.physxWorker.addChild(avatar.activateNodeTwoPtr, avatar.landNodeTwoPtr);
+    physx.physxWorker.addChild(avatar.activateNodeTwoPtr, avatar.activatesNodeSolitaryPtr);
+
+    avatar.holdsNodeSolitaryPtr = avatar.createNode(AnimationNodeType.SOLITARY, 'holdsNodeSolitaryPtr');
+    for (const k in avatar.holdMotionPtro) {
+      const motion = avatar.holdMotionPtro[k];
+      physx.physxWorker.addChild(avatar.holdsNodeSolitaryPtr, motion);
+    }
+    avatar.holdNodeFuncPtr = avatar.createNode(AnimationNodeType.FUNC, 'holdNodeFuncPtr', 0);
+    physx.physxWorker.addChild(avatar.holdNodeFuncPtr, avatar.activateNodeTwoPtr);
+    physx.physxWorker.addChild(avatar.holdNodeFuncPtr, avatar.holdsNodeSolitaryPtr);
+  };
+  createNodes();
 
-    if (avatar.sitState) {
-      return spec => {
-        const {
-          animationTrackName: k,
-          dst,
-          // isTop,
-        } = spec;
+  avatar.rootNodePtr = avatar.holdNodeFuncPtr;
+  physx.physxWorker.setRootNode(avatar.mixerPtr, avatar.rootNodePtr);
+};
 
-        const sitAnimation = sitAnimations[avatar.sitAnimation || defaultSitAnimation];
-        const src2 = sitAnimation.interpolants[k];
-        const v2 = src2.evaluate(1);
+export const _updateAnimation = avatar => {
+  const timeS = performance.now() / 1000;
+
+  const player = metaversefile.getPlayerByAppInstanceId(avatar.app.getComponent('instanceId'));
+
+  const updateValues = () => {
+    const angle = avatar.getAngle();
+    const forwardFactor = 1 - MathUtils.clamp(Math.abs(angle) / (Math.PI / 2), 0, 1);
+    const backwardFactor = 1 - MathUtils.clamp((Math.PI - Math.abs(angle)) / (Math.PI / 2), 0, 1);
+    const leftFactor = 1 - MathUtils.clamp(Math.abs(angle - Math.PI / 2) / (Math.PI / 2), 0, 1);
+    const rightFactor = 1 - MathUtils.clamp(Math.abs(angle - -Math.PI / 2) / (Math.PI / 2), 0, 1);
+    const mirrorFactorReverse = 1 - avatar.mirrorFactor;
+    const mirrorLeftFactor = avatar.mirrorFactor * leftFactor;
+    const mirrorRightFactor = avatar.mirrorFactor * rightFactor;
+    const mirrorLeftFactorReverse = mirrorFactorReverse * leftFactor;
+    const mirrorRightFactorReverse = mirrorFactorReverse * rightFactor;
+
+    physx.physxWorker.setWeight(avatar.walkForwardMotionPtr, forwardFactor);
```

Comment:
I have two concerns with all of these keys:

1) this makes the avatar object bigger, not smaller, so it somewhat defeats the point of doing the animation in WASM
2) this is doing many calls to WASM per frame, per avatar. But it could just be 1.

Instead of this, we can just keep the whole binding vector in a preallocated `ArrayBuffer`, fill it in, and call `.update()` once to absorb it on the WASM side.

Reply:
1. Deleted all `motions/nodes` on JS side, moved to WASM side.

2. `setWeight/Factor/speed` etc all on WASM side, triggered by two js call `updateAvatar` ( for number/floats ) and `updateAvatarString` ( for name/strings ), data passed through preallocated `scratchStack`. ( Should can serialize floats and strings at once, then just need one call, keep improving )